src/Icicle/Avalanche/Statement/Simp/Constructor.hs

```diff
@@ -234,35 +242,75 @@ constructor a_fresh statements
    = Just $ primArrayLength t (primUnpack 0 (ArrayT tx) n)
 
    | (PrimUnsafe (PrimUnsafeArrayIndex tx), [n, aix]) <- prima
-   , Just ts <- tryMeltType tx
-   , tis     <- List.zip ts [0..]
+   , Just tis <- withIndex tryMeltType tx
    = Just $ primPack env tx
    $ fmap (\(t, ix) -> primArrayGet aix t (primUnpack ix (ArrayT tx) n)) tis
 
    | (PrimUpdate (PrimUpdateArrayPut tx), [na, aix, nv]) <- prima
-   , Just ts <- tryMeltType tx
-   , tis     <- List.zip ts [0..]
+   , Just tis <- withIndex tryMeltType tx
    = Just $ primPack env (ArrayT tx)
    $ fmap (\(t, ix) -> primArrayPut aix t (primUnpack ix (ArrayT tx) na)
                                           (primUnpack ix         tx  nv)) tis
 
-   -- comparison
+   -- equality
    | (PrimMinimal (Min.PrimRelation Min.PrimRelationEq t), [nx, ny]) <- prima
-   , Just ts <- tryMeltType t
-   , tis     <- List.zip ts [0..]
-   = Just $ primFold1 primAnd
-   $ fmap (\(tv, i) -> primEq tv (primUnpack i t nx) (primUnpack i t ny)) tis
+   , Just tis <- withIndex tryMeltType t
+   = Just
+   $ primFold1 primAnd
+   $ fmap (withPrim primEq t nx ny) tis
 
    | (PrimMinimal (Min.PrimRelation Min.PrimRelationNe t), [nx, ny]) <- prima
-   , Just ts <- tryMeltType t
-   , tis     <- List.zip ts [0..]
-   = Just $ primFold1 primOr
-   $ fmap (\(tv, i) -> primNe tv (primUnpack i t nx) (primUnpack i t ny)) tis
+   , Just tis <- withIndex tryMeltType t
+   = Just
+   $ primFold1 primOr
+   $ fmap (withPrim primNe t nx ny) tis
```

Comment:
Not sure if it's better, but I guess we could have a function for Op -> AppendOp